src/notation/view/notationviewinputcontroller.cpp

```diff
@@ -657,135 +653,199 @@ void NotationViewInputController::mousePressEvent(QMouseEvent* event)
+            return;
+        }
+    }
+
+    SelectType selectType = SelectType::SINGLE;
+    if (ctx.event->modifiers() & Qt::ShiftModifier) {
+        selectType = SelectType::RANGE;
+    } else if (ctx.event->modifiers() & Qt::ControlModifier) {
+        selectType = SelectType::ADD;
+    } else {
+        if (selection->isRange() && selection->range()->containsItem(ctx.hitElement, ctx.hitStaff)) {
+            return;
+        }
+    }
+
+    viewInteraction()->select({ ctx.hitElement }, selectType, ctx.hitStaff);
+}
+
+void NotationViewInputController::cycleThroughOverlappingHitElements(const std::vector<EngravingItem*>& hitElements,
+                                                                     staff_idx_t hitStaffIndex)
+{
+    size_t numHitElements = hitElements.size();
+    size_t currTop = numHitElements - 1;
+    EngravingItem* e = hitElements[currTop];
+    std::set<EngravingItem*> selectedAtPosition;
+    bool found = false;
+
+    // e is the topmost element in stacking order,
+    // but we want to replace it with "first non-measure element after a selected element"
+    // (if such an element exists)
+    for (size_t i = 0; i <= numHitElements; ++i) {
+        if (found) {
+            e = hitElements[currTop];
+            if (!e->isMeasure()) {
+                break;
+            }
+        } else if (hitElements[currTop]->selected()) {
+            found = true;
+            selectedAtPosition.emplace(hitElements[currTop]);
+            e = nullptr;
+        }
+        currTop = (currTop + 1) % numHitElements;
+    }
+
+    if (e && !e->selected()) {
+        for (EngravingItem* selectedElem : selectedAtPosition) {
+            selectedElem->score()->deselect(selectedElem);
+        }
+        viewInteraction()->select({ e }, SelectType::ADD, hitStaffIndex);
+    }
+}
+
+bool NotationViewInputController::tryHandleClickDragOutgoingRange(const ClickContext& ctx)
+{
+    if (ctx.event->button() != Qt::LeftButton) {
+        return false;
+    }
+
+    if (!ctx.hitElement || ctx.hitElement->isMeasure()) {
+        INotationSelectionPtr selection = viewInteraction()->selection();
+
+        if (selection->isRange() && selection->range()->containsPoint(ctx.logicClickPos)) {
+            m_mouseDownInfo.dragAction = MouseDownInfo::DragOutgoingRange;
+            return true;
+        }
     }
 
     return false;
 }
 
 void NotationViewInputController::handleLeftClick(const ClickContext& ctx)
 {
-    if (ctx.isHitGrip) {
-        viewInteraction()->startEditGrip(ctx.logicClickPos);
+    if (!ctx.hitElement) {
         return;
-    } else {
-        viewInteraction()->endEditGrip();
     }
 
     INotationSelectionPtr selection = viewInteraction()->selection();
 
-    if (!selection->isRange() && ctx.hitElement && ctx.hitElement->needStartEditingAfterSelecting()) {
+    if (!selection->isRange()
+        && ctx.hitElement->selected()
+        && ctx.hitElement->needStartEditingAfterSelecting()) {
```

Comment:
`selection` can be `const`. Also for me these longer/cascading conditions are often worth extracting into a bool:

```suggestion
    const INotationSelectionPtr selection = viewInteraction()->selection();
    
    const bool shouldStartEdit = ctx.hitElement->selected() && ctx.hitElement->needStartEditingAfterSelecting();
    if (!selection->isRange() && shouldStartEdit) {
```